src/localization.rs

```diff
@@ -0,0 +1,361 @@
+//!
+//!
+//! [Fluent]: https://projectfluent.org
+//! [fluent-rs]: https://github.com/projectfluent/fluent-rs
+//! [`LocalizedString`]: struct.LocalizedString.html
+//! [`Env`]: struct.Env.html
+//! [`Data`]: trait.Data.html
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::{fs, io};
+
+use crate::data::Data;
+use crate::env::Env;
+use crate::shell::get_locale;
+
+use fluent_bundle::{FluentArgs, FluentBundle, FluentResource, FluentValue};
+use fluent_locale::{negotiate_languages, NegotiationStrategy};
+use unic_langid::LanguageIdentifier;
+
+// Localization looks for string files in druid/resources, but this path is hardcoded;
+// it will only work if you're running an example from the druid/ directory.
+// At some point we will need to bundle strings with applications, and choose
+// the path dynamically.
+static FALLBACK_STRINGS: &str = include_str!("../resources/i18n/en-US/builtin.ftl");
+
+/// Provides access to the localization strings for the current locale.
+#[allow(dead_code)]
+pub struct L10nManager {
+    res_mgr: ResourceManager,
+    current_bundle: FluentBundle<Arc<FluentResource>>,
+    resources: Vec<String>,
+    current_locale: LanguageIdentifier,
+}
+
+/// Manages a collection of localization files.
+struct ResourceManager {
+    resources: HashMap<String, Arc<FluentResource>>,
+    locales: Vec<LanguageIdentifier>,
+    default_locale: LanguageIdentifier,
+    path_scheme: String,
+}
+
+//NOTE: instead of a closure, at some point we can use something like a lens for this.
+//TODO: this is an Arc so that it can be clone, which is a bound on things like `Menu`.
+/// A closure that generates a localization value.
+type ArgClosure<T> = Arc<dyn Fn(&Env, &T) -> FluentValue<'static> + 'static>;
+
+/// Wraps a closure that generates an argument for localization.
+#[derive(Clone)]
+struct ArgSource<T>(ArgClosure<T>);
+
+/// A string that can be localized based on the current locale.
+///
+/// At it's simplest, a `LocalizedString` is a key that can be resolved
+/// against a map of localized strings for a given locale.
+#[derive(Debug, Clone)]
+pub struct LocalizedString<T> {
+    key: &'static str,
+    placeholder: Option<&'static str>,
+    args: Option<HashMap<&'static str, ArgSource<T>>>,
+    resolved: Option<String>,
+    resolved_lang: Option<LanguageIdentifier>,
+}
+
+//NOTE: much of this is adapted from https://github.com/projectfluent/fluent-rs/blob/master/fluent-resmgr/src/resource_manager.rs
+impl ResourceManager {
+    /// Loads a new localization resource from disk, as needed.
+    fn get_resource(&mut self, res_id: &str, locale: &str) -> Arc<FluentResource> {
+        let path = self
+            .path_scheme
+            .replace("{locale}", locale)
+            .replace("{res_id}", res_id);
+        if let Some(res) = self.resources.get(&path) {
+            res.clone()
+        } else {
+            let string = fs::read_to_string(&path).unwrap_or_else(|_| FALLBACK_STRINGS.to_string());
```

Comment:
The way you wrote the fallback here will only use the fallback in case of a missing file. Much more common scenario where we lazily iterate to fallback is when a string is missing or there are runtime errors during resolution of a string.

That's why we use `Localization` API with cachediterator and return an iterator out of the `L10nManager`.

Reply:
Okay, thanks for pointing that out, I think I was confused. My `L10nManager` is intended to be at about the level of `fluent_fallback::Localization`; it keeps a stack of bundles for the current locale, localizes messages with that stack, and rebuilds that stack when the locale changes. I'll flesh this out a bit more and see how well I'm understanding things.